### ISSUE TITLE
[Proposal] Implement LayerNorm folding for RMS Norm, to cover LLaMA

### ISSUE THREAD

**Issue by neelnanda-io** [neelnanda-io/TransformerLens#372]
### Proposal 

There are two types of LayerNorm: classic LayerNorm that centers to mean zero, scales to standard deviation 1, and then applies learned translation and scaling weights, and RMS norm, which just scales to std 1 and then applies learned scaling weights. Currently from_pretrained only implements classic LayerNorm folding. LLaMA uses RMS Norm which is not covered by this, and it would be nice to do the parts of LayerNorm folding that work for RMS Norm (ie mapping the weights W afterwards to ln.w[:, None] * W)

**Comment by neelnanda-io** [comment]
Note that this should be easy, you basically just need to change some of the flags in `HookedTransformer.load_and_process_state_dict` to allow folding weights for RMS Norm (but not centering!) and edit `HookedTransformer.fold_layer_norm` to only do the bias folding if classic layernorm is used. The model config parameter normalization_type tells you what to use.

**Comment by neelnanda-io** [comment]
Note: Because LLaMA has gated neurons, you'll also need to add support for folding ln2.w into W_gate. This is a bug in the current LN Folding set up, but hasn't triggered because the only models with gated neurons have RMS Norm.

**Comment by neelnanda-io** [comment]
For people who want a temporary work-around, the following code is sufficient

```
for layer in range(n_layers):
    model.blocks[layer].attn.W_K[:] = model.blocks[layer].attn.W_K * model.blocks[layer].ln1.w[None, :, None]
    model.blocks[layer].attn.W_Q[:] = model.blocks[layer].attn.W_Q * model.blocks[layer].ln1.w[None, :, None]
    model.blocks[layer].attn.W_V[:] = model.blocks[layer].attn.W_V * model.blocks[layer].ln1.w[None, :, None]
    model.blocks[layer].ln1.w[:] = torch.ones_like(model.blocks[layer].ln1.w)
    model.blocks[layer].mlp.W_in[:] = model.blocks[layer].mlp.W_in * model.blocks[layer].ln2.w[:, None]
    model.blocks[layer].mlp.W_gate[:] = model.blocks[layer].mlp.W_gate * model.blocks[layer].ln2.w[:, None]
    model.blocks[layer].ln2.w[:] = torch.ones_like(model.blocks[layer].ln2.w)

model.unembed.W_U[:] = model.unembed.W_U * model.ln_final.w[:, None]
model.ln_final.w[:] = torch.ones_like(model.ln_final.w)
```

**Comment by evanhanders** [comment]
Hi everyone! I'll probably work on getting this issue together and into a PR in the next few weeks. If anyone's already working on it or interested in working on it, send me a ping :)

**Comment by collingray** [comment]
@evanhanders Are you still working on this? I'd like to contribute to the repo, so I may take it up if it still needs to be done

**Comment by evanhanders** [comment]
@collingray  -- I got busy and haven't had the time to get around to it! I'm happy to pass it off to you if you're keen to implement it :).

**Comment by collingray** [comment]
@evanhanders Definitely, is there any existing branch or work that I should be aware of before starting on it?

**Comment by evanhanders** [comment]
Nope, no existing branch to start on aside from Neel's workaround above. I talked with @neelnanda-io about this a few weeks ago and he mentioned that it could be neat to add a flag for loading in at half-precision (float16) to keep from blowing out memory, but I can't remember the context of where this would happen in the implementation. And doing so could lead to accuracy issues -- unclear, but could be worth trying and writing a test for.

**Comment by collingray** [comment]
@neelnanda-io good to close?

**Comment by neelnanda-io** [comment]
Closed by #489
